src/Element/PluginSelect.php

```diff
@@ -0,0 +1,119 @@
+use Drupal\Core\Plugin\PluginFormInterface;
+use Drupal\Core\Render\Element\FormElement;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @FormElement("commerce_plugin_select")
+ */
+class PluginSelect extends FormElement {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getInfo() {
+    $class = get_class($this);
+    return [
+      '#input' => TRUE,
+      '#plugin_type' => NULL,
+      '#title' => $this->t('Select plugin'),
+      '#process' => [
+        [$class, 'processPluginSelect'],
+        [$class, 'processAjaxForm'],
+      ],
+    ];
+  }
+
+  /**
+   * Process callback.
+   */
+  public static function processPluginSelect(&$element, FormStateInterface $form_state, &$complete_form) {
+    if (!$element['#plugin_type']) {
+      throw new \InvalidArgumentException('You must specify the plugin type ID.');
+    }
+
+    $element['#tree'] = TRUE;
+
+    /** @var \Drupal\Core\Executable\ExecutableManagerInterface $plugin_manager */
+    $plugin_manager = \Drupal::service('plugin.manager.' . $element['#plugin_type']);
+
+    $values = $element['#value'];
+
+    $target_plugin_id = !empty($values['target_plugin_id']) ? $values['target_plugin_id'] : '_none';
+
+    $ajax_wrapper_id = Html::getUniqueId('ajax-wrapper');
+    $ajax_settings = [
+      'callback' => [get_called_class(), 'pluginFormAjax'],
+      'wrapper' => $ajax_wrapper_id,
+      'options' => [
+        'query' => [
+          'element_parents' => implode('/', $element['#array_parents']),
+        ],
+      ],
+    ];
+
+    // Prefix and suffix used for Ajax replacement.
+    $element['#prefix'] = '<div id="' . $ajax_wrapper_id . '">';
+    $element['#suffix'] = '</div>';
+
+    $element['target_plugin_id'] = [
+      '#type' => 'select',
+      '#title' => $element['#title'],
+      '#multiple' => FALSE,
+      '#options' => [
+        '_none' => t('None'),
+      ],
+      '#ajax' => $ajax_settings,
+      '#default_value' => $target_plugin_id,
+    ];
+
+    $categories = array_combine($element['#categories'], $element['#categories']);
+    foreach ($plugin_manager->getDefinitions() as $definition) {
+      if (isset($categories[$definition['provider']])) {
+        $element['target_plugin_id']['#options'][$definition['id']] = $definition['label'];
+      }
+    }
+
+    if ($target_plugin_id != '_none') {
+      /** @var \Drupal\Core\Executable\ExecutableInterface $plugin */
+      $plugin = $plugin_manager->createInstance($target_plugin_id, $values['target_plugin_configuration']);
+      if ($plugin instanceof  PluginFormInterface) {
+        $element['target_plugin_configuration'] = [
+          '#tree' => TRUE,
+        ] + $plugin->buildConfigurationForm([], $form_state);
+      }
+    }
+
+    return $element;
+  }
+
+  /**
+   * Ajax callback.
+   */
+  public static function pluginFormAjax(&$form, FormStateInterface &$form_state, Request $request) {
+    $form_parents = explode('/', $request->query->get('element_parents'));
```

Comment:
Would it be cleaner to set element_parents as a custom key on the triggering element (the select box)?